### PR TITLE
Add Ipv6Repr and Ipv6Packet to wire

### PR DESCRIPTION
**Pull request by dlrobertson**
- Add the Ipv6Repr and Ipv6Packet structures to the wire module
 - Add basic tests for the new structures
 - Update IpRepr to include the Ipv6 variant

### CODE REVIEW

File: src/wire/ipv6.rs [hunk 1]
```diff
@@ -248,9 +248,310 @@ impl fmt::Display for Cidr {
...
+// |                                                               |
+// +                                                               +
+// |                                                               |
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//
+// See https://tools.ietf.org/html/rfc2460#section-3 for details.
+mod field {
+    use wire::field::*;
+    // 4-bit version number = 6.
+    pub const VERSION:   usize = 0;
```
Comment (whitequark) [inline src/wire/ipv6.rs]: I'd just go with a `VER_TC_FLOW = `, seems like the bit manipulation is less confusing that way.

File: src/wire/ipv6.rs [hunk 2]
```diff
@@ -248,9 +248,310 @@ impl fmt::Display for Cidr {
...
+    /// Return a mutable pointer to the payload.
+    #[inline]
+    pub fn payload_mut(&mut self) -> &mut [u8] {
+        let range = self.header_len()..self.total_len();
+        let data = self.buffer.as_mut();
+        &mut data[range]
+    }
+}
+
+/// A high-level representation of an Internet Protocol version 4 packet header.
```
Comment (whitequark) [inline src/wire/ipv6.rs]: "version 4"

File: src/wire/ipv6.rs [hunk 3]
```diff
@@ -435,4 +736,150 @@ mod test {
...
+    #[test]
+    fn test_total_len_overflow() {
+        let mut bytes = vec![];
+        bytes.extend(&REPR_PACKET_BYTES[..]);
+        Packet::new(&mut bytes).set_payload_len(0x80);
+
+        assert_eq!(Packet::new_checked(&bytes).unwrap_err(),
+                   Error::Truncated);
+    }
+#[test]
```
Comment (whitequark) [inline src/wire/ipv6.rs]: Indentation.

File: src/wire/ip.rs [hunk 4]
```diff
@@ -411,6 +452,8 @@ impl Repr {
         match self {
```
[the remark this answers is not in the document]
Comment (dlrobertson) [reply src/wire/ip.rs]: Should we only change this function or all the `ttl*` functions? E.g. `UdpSocket::ttl`

[the remark this answers is not in the document]
Comment (whitequark) [reply src/wire/ip.rs]: All of them. In general smoltcp aims to be an IPv6-first IPv4-second stack.

[the remark this answers is not in the document]
Comment (dlrobertson) [reply src/wire/ip.rs]: Added a second commit with this in it. I kept the field label as TLL for ipv4, but renamed everything else to hop limit.

File: src/wire/ipv6.rs [hunk 5]
```diff
@@ -248,9 +251,361 @@ impl fmt::Display for Cidr {
     }
 }
 
+/// A read/write wrapper around an Internet Protocol version 6 packet buffer.
+#[derive(Debug, PartialEq)]
+pub struct Packet<T: AsRef<[u8]>> {
+    buffer: T
+}
+
+// Ranges and constants describint the IPv6 header
```
Comment (whitequark) [inline src/wire/ipv6.rs]: describing

File: src/wire/ipv6.rs [hunk 6]
```diff
@@ -248,9 +251,361 @@ impl fmt::Display for Cidr {
...
+    #[inline]
+    pub fn payload_len(&self) -> u16 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u16(&data[field::LENGTH])
+    }
+
+    /// Return the payload length added to the known header length.
+    #[inline]
+    pub fn total_len(&self) -> usize {
+        self.header_len() + (self.payload_len() as usize)
```
Comment (whitequark) [inline src/wire/ipv6.rs]: `as` binds very tight so the parens are not necessary.

File: src/wire/ipv6.rs [hunk 7]
```diff
@@ -248,9 +251,361 @@ impl fmt::Display for Cidr {
...
+impl<'a, T: AsRef<[u8]> + ?Sized> fmt::Display for Packet<&'a T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match Repr::parse(self) {
+            Ok(repr) => write!(f, "{}", repr),
+            Err(err) => {
+                write!(f, "IPv6 ({})", err)?;
+                // Repr::parse has two exceptions, 1) the packet was too short
+                // 2) the packet version number is not 6. If Repr::parse fails
+                // and check_len() returns Ok(()) print the version number.
+                if self.check_len().is_ok() {
```
Comment (whitequark) [inline src/wire/ipv6.rs]: I don't think this branch is necessary, the upper layer will display the packet type. E.g. it's in the ethertype field for Ethernet.

File: src/wire/ipv6.rs [hunk 8]
```diff
@@ -248,9 +251,361 @@ impl fmt::Display for Cidr {
...
+    }
+}
+
+use super::pretty_print::{PrettyPrint, PrettyIndent};
+
+impl<T: AsRef<[u8]>> PrettyPrint for Packet<T> {
+    fn pretty_print(buffer: &AsRef<[u8]>, f: &mut fmt::Formatter,
+                    indent: &mut PrettyIndent) -> fmt::Result {
+        let (ip_repr, payload) = match Packet::new_checked(buffer) {
+            Err(err) => return write!(f, "{}({})\n", indent, err),
```
Comment (whitequark) [inline src/wire/ipv6.rs]: Note that this `\n` is not necessary now.

File: src/wire/ipv6.rs [hunk 9]
```diff
@@ -248,9 +251,361 @@ impl fmt::Display for Cidr {
...
+                    Err(_) => return Ok(()),
+                    Ok(ip_repr) => {
+                        write!(f, "{}{}", indent, ip_repr)?;
+                        (ip_repr, ip_packet.payload())
+                    }
+                }
+            }
+        };
+
+        upper_layer_pretty_print(f, indent, ip_repr, payload)
```
Comment (whitequark) [inline src/wire/ipv6.rs]: Maybe `pretty_print_ip_payload`?

File: src/wire/ipv6.rs [hunk 10]
```diff
@@ -580,6 +580,8 @@ impl fmt::Display for Repr {
 
 use super::pretty_print::{PrettyPrint, PrettyIndent};
 
+// TODO: This is very similar to the implementation for IPv4. Make
+// a way to have less copy and pasted code here.
```
Comment (whitequark) [inline src/wire/ipv6.rs]: In general I'm not too worried about copy/paste in `smoltcp::wire`, various magic to reduce amount of code is an explicit non-goal if you take a look at README.

**Pull request merged**